### DISCUSSION TITLE
How to go about supporting pg_vector postgres extension - supporting the vector type

### DISCUSSION THREAD

**Discussion by alexpusch** [SeaQL/sea-orm#1639] [Q&A] [answered]
Hi, pg_vector postgres extension offers [sqlx support](https://github.com/pgvector/pgvector-rust).

When I try to create a column of the `pgvector::Vector` type the following error occurs

``` rust
#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq)]
#[sea_orm(table_name = "example")]
pub struct Model {
    #[sea_orm(primary_key)]
    pub id: i32,
    // does not help - #[sea_orm(column_type = "custom(\"vector\")")]
    pub vector: pgvector::Vector,
}
```

```
the trait bound `sea_orm::Value: From<Vector>` is not satisfied
```

Is it possible to support types with sqlX support?
Is it possible to add support for external types, or are we limited to the existing variants in the [Value enum](https://docs.rs/sea-orm/latest/sea_orm/enum.Value.html)?

**Comment by billy1624** [comment]
Hey @alexpusch, is it possible to define it as `pub vector: Vec<i32>`?

**Comment by alexpusch** [comment]
@billy1624 trying to just use it with `Vec<f32>` naively fails when using it as binded parameter. The interesting question is - if SeaOrm uses sqlx under the hood. Is it possible to use custom types with [sqlx support](https://github.com/pgvector/pgvector-rust) out of the box?

**Reply by billy1624** [reply]
We will need to update the internal of SeaQuery and SeaORM

**Reply by holmofy** [reply]
How is this issue now?

**Comment by alexpusch** [comment]
@billy1624 I've noticed that some support was added sea-query in https://github.com/SeaQL/sea-query/pull/774
but it seems that sea orm migration still generates the entities as a string with custom type attribute. I can't seem to find a supported Vector type that I manually could replace the String with.

Are there plans to improve vector column support?

```rust
#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq)]
#[sea_orm(table_name = "example")]
pub struct Model {
    #[sea_orm(primary_key)]
    pub id: i32,
    #[sea_orm(column_type = "custom(\"vector\")")]
    pub vector: String,
}
```

**Comment by cosmincatalin** [comment]
I have written a short article on how to use pgVector with SeaORM. Maybe this is useful to people that end up here 🙂 

https://cosminsanda.com/posts/using-pgvector-with-seaorm-in-rust/

**Reply by tyt2y3** [reply]
thank you so much for the article! very well written. I will work on integrating pgvector natively into SeaORM.
we've already added support in SeaQuery https://github.com/SeaQL/sea-query/pull/774

**Reply by tyt2y3** [reply]
Please find this https://github.com/SeaQL/sea-orm/pull/2500

**Comment by tyt2y3** [comment, marked as answer]
Released! https://github.com/SeaQL/sea-orm/releases/tag/1.1.6

https://github.com/SeaQL/sea-orm/blob/master/tests/embedding_tests.rs

**Reply by cosmincatalin** [reply]
Would be nice to demonstrate some pgVector specific operations as well.

**Reply by tyt2y3** [reply]
yeah sad no examples, but there're a few operators https://github.com/SeaQL/sea-query/blob/master/src/extension/postgres/mod.rs#L18

**Reply by cimandef** [reply]
Should `sea-orm-cli` be able to generate the model with the correct types?

**Reply by tyt2y3** [reply]
should be, it's guarded behind a feature flag

try
```
cargo install sea-orm-cli --features=postgres-vector
```

**Reply by holmofy** [reply]
@tyt2y3 Can the bit type of pgvector be supported?  https://github.com/SeaQL/sea-orm/discussions/2675
